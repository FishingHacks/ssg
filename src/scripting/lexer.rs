use std::collections::VecDeque;
use std::ops::ControlFlow;
use std::sync::Arc;

use super::{ByteOffset, ParsingError};

#[derive(Debug)]
enum LexingMode {
    Html,
    Script,
}

#[derive(Debug, Clone, Copy)]
pub enum Token {
    Html(ByteOffset),
    ScriptStart(ByteOffset),
    ScriptEnd(ByteOffset),
    Identifier(ByteOffset),
    StringLiteral(ByteOffset),
    Float(ByteOffset),
    Int(ByteOffset),

    Assign(ByteOffset),
    Equal(ByteOffset),
    NotEqual(ByteOffset),
    GreaterEqual(ByteOffset),
    LesserEqual(ByteOffset),
    Greater(ByteOffset),
    Lesser(ByteOffset),
    Dot(ByteOffset),
    LeftParen(ByteOffset),
    RightParen(ByteOffset),
    Minus(ByteOffset),
    Plus(ByteOffset),
    Mul(ByteOffset),

    Comma(ByteOffset),

    For(ByteOffset),
    In(ByteOffset),
    End(ByteOffset),
    If(ByteOffset),
    Else(ByteOffset),
    Block(ByteOffset),
    Render(ByteOffset),
    Slot(ByteOffset),
    Extend(ByteOffset),
}

impl Token {
    pub fn loc(&self) -> ByteOffset {
        match self {
            Token::Html(byte_offset)
            | Token::ScriptStart(byte_offset)
            | Token::ScriptEnd(byte_offset)
            | Token::Identifier(byte_offset)
            | Token::StringLiteral(byte_offset)
            | Token::Float(byte_offset)
            | Token::Assign(byte_offset)
            | Token::Equal(byte_offset)
            | Token::NotEqual(byte_offset)
            | Token::GreaterEqual(byte_offset)
            | Token::LesserEqual(byte_offset)
            | Token::Greater(byte_offset)
            | Token::Lesser(byte_offset)
            | Token::Dot(byte_offset)
            | Token::LeftParen(byte_offset)
            | Token::RightParen(byte_offset)
            | Token::Minus(byte_offset)
            | Token::Plus(byte_offset)
            | Token::For(byte_offset)
            | Token::In(byte_offset)
            | Token::End(byte_offset)
            | Token::If(byte_offset)
            | Token::Else(byte_offset)
            | Token::Block(byte_offset)
            | Token::Comma(byte_offset)
            | Token::Render(byte_offset)
            | Token::Slot(byte_offset)
            | Token::Mul(byte_offset)
            | Token::Extend(byte_offset)
            | Token::Int(byte_offset) => *byte_offset,
        }
    }

    pub fn is_operator(&self) -> bool {
        matches!(
            self,
            Token::Plus(_)
                | Token::Minus(_)
                | Token::LeftParen(_)
                | Token::RightParen(_)
                | Token::Lesser(_)
                | Token::LesserEqual(_)
                | Token::Greater(_)
                | Token::GreaterEqual(_)
                | Token::NotEqual(_)
                | Token::Equal(_)
                | Token::Dot(_)
                | Token::Mul(_)
        )
    }

    pub fn is_keyword(&self) -> bool {
        matches!(
            self,
            Token::For(_)
                | Token::In(_)
                | Token::If(_)
                | Token::Render(_)
                | Token::Extend(_)
                | Token::Block(_)
                | Token::Slot(_)
                | Token::Else(_)
        )
    }
}

#[derive(Debug)]
pub struct Lexer {
    cursor: usize,
    pub source: Arc<String>,
    mode: LexingMode,
    html_start: Option<usize>,
    script_start: Option<usize>,
    peeked: VecDeque<Result<Token, ParsingError>>,
}

impl Lexer {
    pub fn new(source: Arc<String>) -> Self {
        Self {
            source,
            cursor: 0,
            html_start: None,
            script_start: None,
            peeked: VecDeque::new(),
            mode: LexingMode::Html,
        }
    }

    pub fn peek_n(&mut self, n: usize) -> Option<Result<Token, ParsingError>> {
        let mut buffer = Vec::with_capacity(n);

        while self.peeked.len() <= n {
            let Some(next) = self.next() else {
                break;
            };

            buffer.push(next);
        }

        self.peeked.extend(buffer);

        self.peeked.get(n).cloned()
    }

    pub fn peek(&mut self) -> Option<Result<Token, ParsingError>> {
        self.peek_n(0)
    }

    pub fn is_empty(&mut self) -> bool {
        self.peek().is_none()
    }

    fn is_keyword(&self, ident: &str, offset: ByteOffset) -> Option<Token> {
        match ident {
            "for" => Some(Token::For(offset)),
            "in" => Some(Token::In(offset)),
            "end" => Some(Token::End(offset)),
            "if" => Some(Token::If(offset)),
            "else" => Some(Token::Else(offset)),
            "block" => Some(Token::Block(offset)),
            "render" => Some(Token::Render(offset)),
            "slot" => Some(Token::Slot(offset)),
            "extend" => Some(Token::Extend(offset)),
            _ => None,
        }
    }

    fn advance_script(&mut self, amount: usize) {
        self.cursor += amount;
        self.script_start = Some(self.cursor);
    }

    fn lex_html(&mut self, curr: char, next: Option<char>) -> ControlFlow<Result<Token, ParsingError>, ()> {
        match (curr, next) {
            ('{', Some('{')) => {
                self.mode = LexingMode::Script;
                self.script_start = Some(self.cursor);

                if let Some(start) = self.html_start.take() {
                    return ControlFlow::Break(Ok(Token::Html((start, self.cursor).into())));
                };

                ControlFlow::Continue(())
            }
            // reached the last character, so we return the whole section since
            // next char would be EOF
            (_, None) => {
                self.cursor += 1;
                let start = self.html_start.take().unwrap_or(self.cursor - 1);
                ControlFlow::Break(Ok(Token::Html((start, self.cursor).into())))
            }
            _ => {
                if self.html_start.is_none() {
                    self.html_start = Some(self.cursor);
                }

                self.cursor += 1;

                ControlFlow::Continue(())
            }
        }
    }

    fn lex_script(&mut self, curr: char, next: Option<char>) -> ControlFlow<Result<Token, ParsingError>> {
        match (curr, next) {
            (c, _) if c.is_whitespace() => {
                self.advance_script(c.len_utf8());
                ControlFlow::Continue(())
            }
            ('{', Some('{')) => {
                let start = self
                    .script_start
                    .expect("script start must be set while lexing script content");

                self.advance_script(2);

                ControlFlow::Break(Ok(Token::ScriptStart((start, self.cursor).into())))
            }
            ('}', Some('}')) => {
                self.mode = LexingMode::Html;
                self.cursor += 2;

                let start = self
                    .script_start
                    .take()
                    .expect("script start must be set while lexing script content");

                ControlFlow::Break(Ok(Token::ScriptEnd((start, self.cursor).into())))
            }
            (':', Some('=')) => {
                self.advance_script(2);
                ControlFlow::Break(Ok(Token::Assign((self.cursor - 2, self.cursor).into())))
            }
            ('=', Some('=')) => {
                self.advance_script(2);
                ControlFlow::Break(Ok(Token::Equal((self.cursor - 2, self.cursor).into())))
            }
            ('!', Some('=')) => {
                self.advance_script(2);
                ControlFlow::Break(Ok(Token::NotEqual((self.cursor - 2, self.cursor).into())))
            }
            ('>', Some('=')) => {
                self.advance_script(2);
                ControlFlow::Break(Ok(Token::GreaterEqual((self.cursor - 2, self.cursor).into())))
            }
            ('<', Some('=')) => {
                self.advance_script(2);
                ControlFlow::Break(Ok(Token::LesserEqual((self.cursor - 2, self.cursor).into())))
            }
            ('>', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::Greater((self.cursor - 1, self.cursor).into())))
            }
            ('<', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::Lesser((self.cursor - 1, self.cursor).into())))
            }
            ('.', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::Dot((self.cursor - 1, self.cursor).into())))
            }
            ('(', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::LeftParen((self.cursor - 1, self.cursor).into())))
            }
            (')', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::RightParen((self.cursor - 1, self.cursor).into())))
            }
            ('-', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::Minus((self.cursor - 1, self.cursor).into())))
            }
            ('*', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::Mul((self.cursor - 1, self.cursor).into())))
            }
            (',', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::Comma((self.cursor - 1, self.cursor).into())))
            }
            ('+', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::Plus((self.cursor - 1, self.cursor).into())))
            }
            ('"' | '\'', _) => {
                self.cursor += 1;
                let start = self.cursor;
                let source = &self.source[self.cursor..];

                let Some(length) = source.find(curr) else {
                    return ControlFlow::Break(Err(ParsingError {
                        src: self.source.clone(),
                        at: (start..self.source.len()).into(),
                        help: "unterminated string".into(),
                    }));
                };

                // skip past the closing " for the next iterations but don't
                // include it on the offset range
                self.advance_script(length + 1);
                let end = start + length;

                ControlFlow::Break(Ok(Token::StringLiteral((start, end).into())))
            }
            ('0'..='9', _) => {
                let start = self.cursor;
                let source = &self.source[self.cursor..];

                let length = source
                    .find(|c| !matches!(c, '_' | '.' | '0'..='9'))
                    .unwrap_or(self.source.len());

                self.advance_script(length);
                let end = start + length;
                let literal = &self.source[start..end];

                let mut dotted = literal.splitn(3, '.');
                match (dotted.next(), dotted.next(), dotted.next()) {
                    // found at least 2 dots on a number. thats a invalid float
                    (Some(_), Some(_), Some(_)) => {
                        return ControlFlow::Break(Err(ParsingError {
                            src: self.source.clone(),
                            at: (start..end).into(),
                            help: "invalid numeral".into(),
                        }));
                    }
                    _ => {
                        // literal is either a integer or a valid floating point
                        // so skip doing anything here
                    }
                };

                let is_float = literal.contains('.');
                if is_float {
                    return ControlFlow::Break(Ok(Token::Float((start, end).into())));
                }

                ControlFlow::Break(Ok(Token::Int((start, end).into())))
            }
            (c, _) => {
                if c.is_identifier_starter() {
                    let start = self.cursor;
                    let source = &self.source[self.cursor..];

                    let length = source
                        .find(|c: char| !c.is_valid_identifier())
                        .unwrap_or(self.source.len());

                    self.advance_script(length);

                    let end = start + length;
                    let content = &self.source[start..end];
                    let offset = (start, end).into();

                    if let Some(keyword) = self.is_keyword(content, offset) {
                        return ControlFlow::Break(Ok(keyword));
                    }

                    return ControlFlow::Break(Ok(Token::Identifier(offset)));
                }

                println!("{c:?}");
                panic!();
            }
        }
    }
}

impl Iterator for Lexer {
    type Item = Result<Token, ParsingError>;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            if let Some(peeked) = self.peeked.pop_front() {
                return Some(peeked);
            }

            let source = &self.source[self.cursor..];
            let mut chars = source.chars().peekable();

            let ch = chars.next()?;
            let next = chars.peek().copied();

            match self.mode {
                LexingMode::Html => match self.lex_html(ch, next) {
                    ControlFlow::Continue(_) => continue,
                    ControlFlow::Break(token) => return Some(token),
                },
                LexingMode::Script => match self.lex_script(ch, next) {
                    ControlFlow::Continue(_) => continue,
                    ControlFlow::Break(token) => return Some(token),
                },
            }
        }
    }
}

trait Identifier {
    fn is_identifier_starter(&self) -> bool;
    fn is_valid_identifier(&self) -> bool;
}

impl Identifier for char {
    fn is_identifier_starter(&self) -> bool {
        self.is_alphabetic() || matches!(self, '_' | '$')
    }

    fn is_valid_identifier(&self) -> bool {
        self.is_ascii_alphanumeric() || matches!(self, '_')
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // allowing dead code as this is simply a test helper
    #[derive(Debug)]
    #[allow(dead_code)]
    struct TokenSnapshot<'snap> {
        token: Token,
        content: &'snap str,
    }

    impl<'snap> TokenSnapshot<'snap> {
        fn with_content(token: Token, content: &'snap str) -> Self {
            Self { content, token }
        }
    }

    #[test]
    fn test_lexing_simple_document() {
        let source = [
            "<!DOCTYPE html>",
            "<html lang=\"en\">",
            "  <head id=\"head\">",
            "    <meta charset=\"UTF-8\">",
            "    <meta name=\"description\" content=\"{{ page.description }}\">",
            "",
            "    {{ if page.author }}",
            "    <title>{{ page.title }} - {{ page.author }}</title>",
            "    {{ else }}",
            "    <title>{{ page.title }}</title>",
            "    {{ end }}",
            "",
            "    {{ slot \"head\" }}",
            "  </head>",
            "  <body>",
            "    <main>",
            "      {{ slot \"page\" }}",
            "    </main>",
            "  </body>",
            "</html>",
        ]
        .join("\n");

        let source = Arc::new(source);
        let lexer = Lexer::new(source.clone());

        let tokens = lexer
            .map(|t| t.unwrap())
            .map(|token| TokenSnapshot::with_content(token, &source[token.loc().range()]))
            .collect::<Vec<_>>();

        insta::assert_debug_snapshot!(tokens);
    }

    #[test]
    fn test_lexing_language_features() {
        // this is purposefully nonsensical as it is supposed to test lexing of
        // language features.
        let source = [
            // if statement
            "{{ if page.author }}",
            "<title>{{ page.title }} - {{ page.author }}</title>",
            "{{ else }}",
            "<title>{{ page.title }}</title>",
            "{{ end }}",
            "",
            // for loop
            "{{ for tag in page.tags }}",
            "<span>{{ tag.name }}</span>",
            "{{ end }}",
            "",
            // slots
            "{{ slot \"head\" }}",
            "{{ slot \"page\" }}",
            "",
            // blocks
            "{{ block \"head\" }}",
            "{{ block \"page\" }}",
            // render
            "{{ render \"something.html\" }}",
            "{{ render \"another_thing.html\" }}",
            // extend
            "{{ extend \"template.html\" }}",
            // calling functions
            "{{ site.assets.get('asset_name.extension') }}",
            // variable
            "{{ name := page.title }}",
            // comparisons
            "{{ page.title == \"some title\" }}",
            "{{ page.title != \"some title\" }}",
            "{{ page.words > 1337 }}",
            "{{ page.words >= 1337 }}",
            "{{ page.words < 1337 }}",
            "{{ page.words <= 1337 }}",
            "{{ page.words != 1337 }}",
            // numbers
            "{{ 1337 }}",
            "{{ -1337 }}",
            "{{ 13.37 }}",
            "{{ -13.37 }}",
            "{{ -1337 }}",
            "{{ 13.37 }}",
            "{{ -13.37 }}",
            "{{ 1337 - -1337 }}",
            "{{ -1337 + -1337 }}",
            "{{ -1337+-1337 }}",
            "{{ -1337.1337+-13.33 }}",
        ]
        .join("\n");

        let source = Arc::new(source);
        let lexer = Lexer::new(source.clone());

        let tokens = lexer
            .map(|t| t.unwrap())
            .map(|token| TokenSnapshot::with_content(token, &source[token.loc().range()]))
            .collect::<Vec<_>>();

        insta::assert_debug_snapshot!(tokens);
    }
}
