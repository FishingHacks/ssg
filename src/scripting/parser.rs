use std::sync::Arc;

use miette::SourceSpan;

use super::lexer::{Lexer, Token};
use super::{Ast, ByteOffset, Expr, Operator, ParsingError};
use crate::scripting::AstNode;

const EXPECTED_STR: &str = "Expected a string";
const EXPECTED_IDENT: &str = "Expected an identifier";
const EXPECTED_LIT_OR_PARENS: &str = "Expected (, an identifier or a literal";
const INVALID_INT: &str = "That is an invalid integer";
const INVALID_FLOAT: &str = "That is an invalid float";

mod precedences {
    pub const BASE: u8 = 0;
    pub const SUM: u8 = 3;
    pub const MUL: u8 = 4;
    pub const ASSOC: u8 = 5;
    pub const APPLY: u8 = 6;
}

fn get_precedence(token: Token) -> u8 {
    match token {
        Token::Plus(_) => precedences::SUM,
        Token::Mul(_) => precedences::MUL,
        Token::LeftParen(_) => precedences::APPLY,
        _ => precedences::BASE,
    }
}

#[macro_export]
macro_rules! peek {
    ($lexer:expr) => {
        $lexer.peek().transpose()?
    };

    ($lexer:expr, $amount:expr) => {
        $lexer.peek_n($amount).transpose()?
    };
}

#[macro_export]
macro_rules! peek_bail {
    ($lexer:expr) => {{
        let Some(token) = peek!($lexer) else {
            return Err(ParsingError {
                src: $lexer.source.clone(),
                at: (0..1).into(),
                help: "Syntax error: unterminated expression".into(),
            });
        };

        token
    }};

    ($lexer:expr, $amount:expr) => {{
        let Some(token) = peek!($lexer, $amount) else {
            return Err(ParsingError {
                src: $lexer.source.clone(),
                at: (0..1).into(),
                help: "Syntax error: unterminated expression".into(),
            });
        };

        token
    }};
}

#[macro_export]
macro_rules! consume {
    ($lexer:expr) => {
        $lexer.next().transpose()?
    };

    ($lexer:expr, $amount:expr) => {{
        for _ in 0..$amount {
            $lexer.next().transpose()?;
        }
    }};
}

#[macro_export]
macro_rules! expect {
    ($lexer:expr, $expected:pat) => {{
        let Some(token) = $lexer.next().transpose()? else {
            return Err(ParsingError {
                at: (0..1).into(),
                help: format!("Unexpected token: expected {:?} but found EOF", stringify!($expected)),
                src: $lexer.source.clone(),
            });
        };

        if matches!(token, $expected) {
            Ok(token)
        } else {
            Err(ParsingError {
                at: token.loc().range().into(),
                help: format!(
                    "Unexpecte token: expected {:?} but found {token:?}",
                    stringify!($expected)
                ),
                src: $lexer.source.clone(),
            })
        }
    }};
}

#[derive(Debug)]
pub struct Parser {
    source: Arc<String>,
    lexer: Lexer,
}

impl Parser {
    pub fn new(source: Arc<String>, lexer: Lexer) -> Self {
        Self { source, lexer }
    }

    pub fn parse_all(&mut self) -> Result<Ast, ParsingError> {
        let mut nodes = vec![];

        while !self.lexer.is_empty() {
            let node = self.parse()?;
            nodes.push(node);
        }

        Ok(Ast {
            source: self.source.clone(),
            nodes,
        })
    }

    fn parse(&mut self) -> Result<AstNode, ParsingError> {
        // token will exist as we only loop while lexer is not empty
        let token = peek!(self.lexer).unwrap();

        match token {
            Token::Html(offset) => {
                consume!(self.lexer);
                Ok(AstNode::Html(offset))
            }
            Token::ScriptStart(_) => self.parse_script_block(),
            t => unreachable!("{t:?}"),
        }
    }

    fn parse_script_block(&mut self) -> Result<AstNode, ParsingError> {
        expect!(self.lexer, Token::ScriptStart(_))?;

        let token = peek_bail!(self.lexer);

        if matches!(token, Token::For(_)) {
            let keyword = expect!(self.lexer, Token::For(_))?;
            return self.parse_for_loop(keyword);
        }

        if matches!(token, Token::Render(_)) {
            let keyword = expect!(self.lexer, Token::Render(_))?;
            let component = expect!(self.lexer, Token::StringLiteral(_))?.loc();
            let offset = keyword.loc() + component;
            expect!(self.lexer, Token::ScriptEnd(_))?;
            return Ok(AstNode::Render { component, offset });
        }

        if matches!(token, Token::Extend(_)) {
            let keyword = expect!(self.lexer, Token::Extend(_))?;
            let template = expect!(self.lexer, Token::StringLiteral(_))?.loc();
            let offset = keyword.loc() + template;
            expect!(self.lexer, Token::ScriptEnd(_))?;
            return Ok(AstNode::Extend { template, offset });
        }

        if matches!(token, Token::Block(_)) {
            let keyword = expect!(self.lexer, Token::Block(_))?;
            let name = expect!(self.lexer, Token::StringLiteral(_))?.loc();
            let offset = keyword.loc() + name;
            expect!(self.lexer, Token::ScriptEnd(_))?;
            return Ok(AstNode::Block { name, offset });
        }

        let expr = self.parse_expression(precedences::BASE)?;

        expect!(self.lexer, Token::ScriptEnd(_))?;
        Ok(expr)
    }

    fn parse_expression(&mut self, min_precedence: u8) -> Result<AstNode, ParsingError> {
        let left = self.lexer.next().transpose()?;
        let mut left = match left {
            Some(Token::Identifier(offset)) => AstNode::Identifier(offset),
            Some(Token::Int(offset)) => AstNode::IntLiteral(offset),
            Some(Token::Float(offset)) => AstNode::FloatLiteral(offset),
            Some(Token::StringLiteral(offset)) => AstNode::StringLiteral(offset),
            Some(token) if token.is_operator() => self.parse_operator(token)?,
            None => unreachable!(),
            t => unreachable!("{t:?}"),
        };

        if let Token::Assign(_) = peek_bail!(self.lexer) {
            expect!(self.lexer, Token::Assign(_))?;

            let value = self.parse_expression(precedences::BASE)?;

            return Ok(AstNode::Variable {
                ident: left.loc(),
                value: Box::new(value),
            });
        };

        loop {
            let Some(next) = peek!(self.lexer) else {
                return Ok(left);
            };

            if !next.is_operator() {
                return Ok(left);
            }

            let precedence = get_precedence(next);

            if precedence <= min_precedence {
                break;
            }

            let Some(operator) = consume!(self.lexer) else {
                unreachable!();
            };

            let operator = Operator::new_unchecked(operator);

            let right = self.parse_expression(precedence)?;

            left = AstNode::BinaryOp {
                offset: left.loc() + right.loc(),
                left: Box::new(left),
                right: Box::new(right),
                operator,
            }
        }

        Ok(left)
    }

    fn parse_operator(&mut self, token: Token) -> Result<AstNode, ParsingError> {
        let operator = Operator::new_unchecked(token);

        match operator {
            Operator::LeftParen => {
                let left = self.parse_expression(precedences::BASE)?;
                expect!(self.lexer, Token::RightParen(_))?;
                Ok(left)
            }
            _ => unreachable!(),
        }
    }

    fn parse_for_loop(&mut self, keyword: Token) -> Result<AstNode, ParsingError> {
        let identifier = expect!(self.lexer, Token::Identifier(_))?.loc();

        let index = match peek!(self.lexer) {
            Some(Token::Comma(_)) => {
                consume!(self.lexer);
                Some(expect!(self.lexer, Token::Identifier(_))?.loc())
            }
            _ => None,
        };

        expect!(self.lexer, Token::In(_))?;
        let list = expect!(self.lexer, Token::Identifier(_))?.loc();
        let offset = keyword.loc() + list;

        expect!(self.lexer, Token::ScriptEnd(_))?;

        let body = self.parse_until_end_block()?;

        Ok(AstNode::ForLoop {
            identifier,
            index,
            list,
            offset,
            body,
        })
    }

    fn parse_until_end_block(&mut self) -> Result<Vec<AstNode>, ParsingError> {
        let mut body = vec![];

        loop {
            let first = peek_bail!(self.lexer);
            let second = peek_bail!(self.lexer, 1);

            if matches!(first, Token::ScriptStart(_)) && matches!(second, Token::End(_)) {
                consume!(self.lexer, 2);
                expect!(self.lexer, Token::ScriptEnd(_))?;
                break;
            }

            let node = self.parse()?;

            body.push(node);
        }

        Ok(body)
    }
}

#[cfg(test)]
mod tests {

    use super::*;

    #[allow(dead_code)]
    #[derive(Debug)]
    enum AstNodeSnapshot<'ast> {
        Html {
            offset: ByteOffset,
            content: &'ast str,
        },
        Identifier {
            offset: ByteOffset,
            content: &'ast str,
        },
        IntLiteral {
            offset: ByteOffset,
            content: &'ast str,
        },
        FloatLiteral {
            offset: ByteOffset,
            content: &'ast str,
        },
        StringLiteral {
            offset: ByteOffset,
            content: &'ast str,
        },
        Variable {
            ident: ByteOffset,
            value: Box<AstNodeSnapshot<'ast>>,
            content: &'ast str,
        },
        BinaryOp {
            left: Box<AstNodeSnapshot<'ast>>,
            right: Box<AstNodeSnapshot<'ast>>,
            offset: ByteOffset,
            content: &'ast str,
            operator: Operator,
        },
        ForLoop {
            identifier: ByteOffset,
            index: Option<ByteOffset>,
            list: ByteOffset,
            offset: ByteOffset,
            content: &'ast str,
            body: Vec<AstNodeSnapshot<'ast>>,
        },
        Render {
            component: ByteOffset,
            offset: ByteOffset,
            content: &'ast str,
        },
        Extend {
            template: ByteOffset,
            offset: ByteOffset,
            content: &'ast str,
        },
        Block {
            name: ByteOffset,
            offset: ByteOffset,
            content: &'ast str,
        },
    }

    fn node_to_snapshot(node: AstNode, source: &str) -> AstNodeSnapshot<'_> {
        match node {
            AstNode::Html(offset) => AstNodeSnapshot::Html {
                content: &source[offset.range()],
                offset,
            },
            AstNode::Identifier(offset) => AstNodeSnapshot::Identifier {
                content: &source[offset.range()],
                offset,
            },
            AstNode::IntLiteral(offset) => AstNodeSnapshot::IntLiteral {
                content: &source[offset.range()],
                offset,
            },
            AstNode::FloatLiteral(offset) => AstNodeSnapshot::FloatLiteral {
                content: &source[offset.range()],
                offset,
            },
            AstNode::StringLiteral(offset) => AstNodeSnapshot::StringLiteral {
                content: &source[offset.range()],
                offset,
            },
            AstNode::Variable { ident, value } => AstNodeSnapshot::Variable {
                content: &source[(ident + value.loc()).range()],
                ident,
                value: Box::new(node_to_snapshot(*value, source)),
            },
            AstNode::BinaryOp {
                left,
                right,
                operator,
                offset,
            } => AstNodeSnapshot::BinaryOp {
                content: &source[offset.range()],
                left: Box::new(node_to_snapshot(*left, source)),
                right: Box::new(node_to_snapshot(*right, source)),
                operator,
                offset,
            },
            AstNode::ForLoop {
                identifier,
                index,
                list,
                offset,
                body,
            } => AstNodeSnapshot::ForLoop {
                content: &source[offset.range()],
                identifier,
                offset,
                index,
                list,
                body: body.into_iter().map(|node| node_to_snapshot(node, source)).collect(),
            },
            AstNode::Render { component, offset } => AstNodeSnapshot::Render {
                content: &source[offset.range()],
                component,
                offset,
            },
            AstNode::Extend { template, offset } => AstNodeSnapshot::Extend {
                content: &source[offset.range()],
                template,
                offset,
            },
            AstNode::Block { name, offset } => AstNodeSnapshot::Block {
                content: &source[offset.range()],
                name,
                offset,
            },
        }
    }

    #[test]
    fn test_parse_identifier() {
        let source = Arc::new(String::from("{{ var_name }}"));

        let lexer = Lexer::new(source.clone());
        let mut parser = Parser::new(source.clone(), lexer);
        let ast = parser.parse_all().unwrap();

        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn test_parse_var_identifier() {
        let source = Arc::new(String::from("{{ var_name := something_else }}"));

        let lexer = Lexer::new(source.clone());
        let mut parser = Parser::new(source.clone(), lexer);
        let ast = parser.parse_all().unwrap();

        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn test_parse_var_number() {
        let source = Arc::new(String::from("{{ var_name := 1337 }}"));

        let lexer = Lexer::new(source.clone());
        let mut parser = Parser::new(source.clone(), lexer);
        let ast = parser.parse_all().unwrap();

        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn test_parse_var_float() {
        let source = Arc::new(String::from("{{ var_name := 13.37 }}"));

        let lexer = Lexer::new(source.clone());
        let mut parser = Parser::new(source.clone(), lexer);
        let ast = parser.parse_all().unwrap();

        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn test_parse_var_string() {
        let source = Arc::new(String::from("{{ var_name := \"hello\" }}"));

        let lexer = Lexer::new(source.clone());
        let mut parser = Parser::new(source.clone(), lexer);
        let ast = parser.parse_all().unwrap();

        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn test_parse_var_expr() {
        let source = Arc::new(String::from("{{ var_name := (100 + 300) * 200 }}"));

        let lexer = Lexer::new(source.clone());
        let mut parser = Parser::new(source.clone(), lexer);
        let ast = parser.parse_all().unwrap();

        let ast = ast
            .nodes
            .into_iter()
            .map(|node| node_to_snapshot(node, &source))
            .collect::<Vec<_>>();

        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn test_parse_for_loop() {
        let source = [
            "{{ for tag, idx in tags }}",
            "<div>",
            "  <span>Hello World!</span>",
            "</div>",
            "{{ render \"button.html\" }}",
            "<div>",
            "  <span>Hello Another</span>",
            "</div>",
            "{{ end }}",
        ]
        .join("\n");

        let source = Arc::new(source);
        let lexer = Lexer::new(source.clone());
        let mut parser = Parser::new(source.clone(), lexer);
        let ast = parser.parse_all().unwrap();

        let ast = ast
            .nodes
            .into_iter()
            .map(|node| node_to_snapshot(node, &source))
            .collect::<Vec<_>>();

        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn test_parse_for_without_idx() {
        let source = [
            "{{ for tag in tags }}",
            "<div>",
            "  <span>Hello World!</span>",
            "</div>",
            "{{ render \"button.html\" }}",
            "<div>",
            "  <span>Hello Another</span>",
            "</div>",
            "{{ end }}",
        ]
        .join("\n");

        let source = Arc::new(source);
        let lexer = Lexer::new(source.clone());
        let mut parser = Parser::new(source.clone(), lexer);
        let ast = parser.parse_all().unwrap();

        let ast = ast
            .nodes
            .into_iter()
            .map(|node| node_to_snapshot(node, &source))
            .collect::<Vec<_>>();

        insta::assert_debug_snapshot!(ast);
    }
}
