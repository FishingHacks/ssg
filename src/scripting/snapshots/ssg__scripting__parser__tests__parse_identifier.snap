---
source: src/scripting/parser.rs
expression: ast
---
Ast {
    source: "{{ var_name }}",
    nodes: [
        Identifier(
            ByteOffset {
                start: 3,
                end: 11,
            },
        ),
    ],
}
